### PR TITLE
Add validating webhook for shoots

### PR DESCRIPTION
**Pull request by timuthy**
**What this PR does / why we need it**:
This PR adds a validating webhook that is responsible for checking resources of the group `core.gardener.cloud` (at the moment only `shoot` resources).

With [GEP-4](https://github.com/gardener/gardener/blob/master/docs/proposals/04-new-core-gardener-cloud-apis.md) cloud specific information was externalized and as a consequence cannot be validated by the Gardener anymore. On the other hand, it should be prevented that invalid `shoot` objects are created or updated.

**Release note**:

```noteworthy operator
The `gardener-extension-provider-openstack` now offers a validating webhook which checks shoot resources of type `openstack` in the `core.gardener.cloud` group. Operators should register this webhook in the Garden cluster to further prevent invalid modifications on `openstack` shoots.
```

### CODE REVIEW

File: charts/gardener-extension-validator-openstack/charts/runtime/templates/deployment.yaml [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+spec:
+  revisionHistoryLimit: 0
+  replicas: {{ .Values.global.replicaCount }}
+  selector:
+    matchLabels:
+{{ include "labels" . | indent 6 }}
+  template:
+    metadata:
+      annotations:
+        checksum/secret-gardener-extensions-validator-openstack-cert: {{ include (print $.Template.BasePath "/secret-cert.yaml") . | sha256sum }}
```
Comment (rfranzke) [inline charts/gardener-extension-validator-openstack/charts/runtime/templates/deployment.yaml]: s/extensions/extension everywhere

File: charts/gardener-extension-validator-openstack/charts/runtime/templates/secret-cert.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: gardener-extensions-validator-openstack-cert
```
Comment (rfranzke) [inline charts/gardener-extension-validator-openstack/charts/runtime/templates/secret-cert.yaml]:
```suggestion
  name: gardener-extension-validator-openstack-cert
```

File: charts/gardener-extension-validator-openstack/charts/runtime/templates/secret-kubeconfig.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
+{{- if .Values.global.kubeconfig }}
+apiVersion: v1
+kind: Secret
+metadata:
+  name: gardener-extensions-validator-openstack-kubeconfig
```
Comment (rfranzke) [inline charts/gardener-extension-validator-openstack/charts/runtime/templates/secret-kubeconfig.yaml]:
```suggestion
  name: gardener-extension-validator-openstack-kubeconfig
```

File: charts/gardener-extension-validator-openstack/charts/runtime/values.yaml [hunk 4]
```diff
@@ -0,0 +1,26 @@
+global:
```
Comment (rfranzke) [inline charts/gardener-extension-validator-openstack/charts/runtime/values.yaml]: symlink to `../../values.yaml`

File: charts/gardener-extension-validator-openstack/templates/_helpers.tpl [hunk 5]
```diff
@@ -0,0 +1,23 @@
+{{- define "name" -}}
+gardener-extensions-validator-openstack
```
Comment (rfranzke) [inline charts/gardener-extension-validator-openstack/templates/_helpers.tpl]:
```suggestion
gardener-extension-validator-openstack
```

File: example/40-validatingwebhookconfiguration.yaml [hunk 6]
```diff
@@ -0,0 +1,23 @@
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingWebhookConfiguration
+metadata:
+  name: gardener-extensions-validator-openstack
```
Comment (rfranzke) [inline example/40-validatingwebhookconfiguration.yaml]:
```suggestion
  name: gardener-extension-validator-openstack
```

File: example/validator-openstack-certs/tls.crt [hunk 7]
```diff
@@ -0,0 +1,23 @@
+-----BEGIN CERTIFICATE-----
```
Comment (rfranzke) [inline example/validator-openstack-certs/tls.crt]: "Provider AWS" inside which might be confusing

File: pkg/apis/openstack/validation/shoot.go [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+// ValidateWorkers validates the workers of a Shoot.
+func ValidateWorkers(workers []core.Worker, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	for i, worker := range workers {
+
+		if len(worker.Zones) == 0 {
+			allErrs = append(allErrs, field.Required(fldPath.Index(i).Child("zones"), "at least one zone must be configured for zoned clusters"))
```
Comment (rfranzke) [inline pkg/apis/openstack/validation/shoot.go]:
```suggestion
			allErrs = append(allErrs, field.Required(fldPath.Index(i).Child("zones"), "at least one zone must be configured"))
```

Comment (timuthy) [reply pkg/apis/openstack/validation/shoot.go]: nice catch

File: pkg/validator/shoot_validator.go [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+
+	if errList := openstackvalidation.ValidateWorkersUpdate(oldShoot.Spec.Provider.Workers, shoot.Spec.Provider.Workers, workersPath); len(errList) > 0 {
+		return errList.ToAggregate()
+	}
+
+	return v.validateShoot(ctx, valContext)
+}
+
+func (v *Shoot) validateShoot(ctx context.Context, context *validationContext) error {
+	if errList := openstackvalidation.ValidateNetworking(context.shoot.Spec.Networking, nwPath); len(errList) != 0 {
```
Comment (tim-ebert) [inline pkg/validator/shoot_validator.go]: Is there a specific reason, why you don't collect the errors from the different validation functions to return a combined error list?
If not I would suggest appending the returned errors to a `field.ErrorList` and returning it at the end of the function :)

File: example/40-validatingwebhookconfiguration.yaml [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+    operations:
+    - CREATE
+    - UPDATE
+    resources:
+    - shoots
+  failurePolicy: Fail
+  objectSelector: {}
+  namespaceSelector: {}
+  clientConfig:
+    url: "https://localhost:9443/webhooks/validate-shoots"
```
Comment (MartinWeindel) [inline example/40-validatingwebhookconfiguration.yaml]: inconsistent with webhook registration path `webhooks/validate-shoot`, see cmd/gardener-extension-validator-openstack/app/app.go, line 79

**Comment by timuthy** [comment]
Thanks for your review. I applied your suggestions, PTAL.

File: example/40-validatingwebhookconfiguration.yaml (hunk 6 above)
Comment (rfranzke) [reply example/40-validatingwebhookconfiguration.yaml]: ping, not yet resolved

Comment (timuthy) [reply example/40-validatingwebhookconfiguration.yaml]: Sorry, I must have missed this one.

**Comment by timuthy** [comment]
Thank you, I applied your suggestions. PTAL.

**Pull request merged**